From 5b8757a1bd5ecc60d30b539bb7eaabd5a410aea5 Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@gmail.com>
Date: Sat, 22 May 2021 23:31:43 -0700
Subject: [PATCH 108/196] Add `enableMinerFund` to config to enable testing
 (#184)

This commit adds `enableMinerFund` to the config so that it can be
overriden in w/o setting gArgs. It also adds flags to the
`BlockAssembler` options as well as the validation options in order to
support passing these configs down to the appropriate code. However,
current testing setup relies on the global `GetConfig()` singleton. This
singleton always returns the chainparams from the global `Params()`
singleton function. The testing setup changes the Params after
initialization, thus we need to ensure that `EnableMinerFund()` returns
the appropriate default consensus parameter if it has not been formally
set by the command line or the testing setup. In the future, global
singletons should be removed so that tests can be ran in parallel.
However, this requires substantial refactors to the entire codebase to
ensure proper initialization ordering and encapsulation of
functionality.
---
 src/config.cpp                       | 28 ++++++++++--
 src/config.h                         | 18 ++++++++
 src/init.cpp                         |  5 +++
 src/miner.cpp                        |  8 +++-
 src/miner.h                          |  2 +
 src/minerfund.cpp                    |  4 +-
 src/minerfund.h                      |  1 +
 src/rpc/mining.cpp                   |  5 ++-
 src/test/blockfilter_index_tests.cpp |  2 +-
 src/test/miner_tests.cpp             | 66 ++++++++++++++--------------
 src/validation.cpp                   |  5 ++-
 src/validation.h                     | 16 +++++--
 12 files changed, 112 insertions(+), 48 deletions(-)

diff --git a/src/config.cpp b/src/config.cpp
index cb2e8bd70..927f29f80 100644
--- a/src/config.cpp
+++ b/src/config.cpp
@@ -8,7 +8,8 @@
 #include <consensus/consensus.h> // DEFAULT_MAX_BLOCK_SIZE
 
 GlobalConfig::GlobalConfig()
-    : useCashAddr(false), nMaxBlockSize(DEFAULT_MAX_BLOCK_SIZE) {}
+    : useCashAddr(false), enableMinerFund(MinerFundStatus::Unset),
+      nMaxBlockSize(DEFAULT_MAX_BLOCK_SIZE) {}
 
 bool GlobalConfig::SetMaxBlockSize(uint64_t maxBlockSize) {
     // Do not allow maxBlockSize to be set below historic 1MB limit
@@ -42,14 +43,26 @@ bool GlobalConfig::UseCashAddrEncoding() const {
     return useCashAddr;
 }
 
+void GlobalConfig::SetEnableMinerFund(bool c) {
+    enableMinerFund = c ? MinerFundStatus::Enabled : MinerFundStatus::Disabled;
+}
+bool GlobalConfig::EnableMinerFund() const {
+    return enableMinerFund == MinerFundStatus::Unset
+               ? GetChainParams().GetConsensus().enableMinerFund
+               : enableMinerFund == MinerFundStatus::Enabled;
+}
+
 DummyConfig::DummyConfig()
-    : chainParams(CreateChainParams(CBaseChainParams::REGTEST)) {}
+    : chainParams(CreateChainParams(CBaseChainParams::REGTEST)),
+      enableMinerFund(chainParams->GetConsensus().enableMinerFund) {}
 
 DummyConfig::DummyConfig(std::string net)
-    : chainParams(CreateChainParams(net)) {}
+    : chainParams(CreateChainParams(net)),
+      enableMinerFund(chainParams->GetConsensus().enableMinerFund) {}
 
 DummyConfig::DummyConfig(std::unique_ptr<CChainParams> chainParamsIn)
-    : chainParams(std::move(chainParamsIn)) {}
+    : chainParams(std::move(chainParamsIn)),
+      enableMinerFund(chainParams->GetConsensus().enableMinerFund) {}
 
 void DummyConfig::SetChainParams(std::string net) {
     chainParams = CreateChainParams(net);
@@ -62,3 +75,10 @@ void GlobalConfig::SetExcessUTXOCharge(Amount fee) {
 Amount GlobalConfig::GetExcessUTXOCharge() const {
     return excessUTXOCharge;
 }
+
+void DummyConfig::SetEnableMinerFund(bool c) {
+    enableMinerFund = c;
+}
+bool DummyConfig::EnableMinerFund() const {
+    return enableMinerFund;
+}
diff --git a/src/config.h b/src/config.h
index 56fbe7dfa..dd3b66fcb 100644
--- a/src/config.h
+++ b/src/config.h
@@ -26,9 +26,19 @@ public:
 
     virtual void SetExcessUTXOCharge(Amount amt) = 0;
     virtual Amount GetExcessUTXOCharge() const = 0;
+
+    virtual void SetEnableMinerFund(bool) = 0;
+    virtual bool EnableMinerFund() const = 0;
 };
 
 class GlobalConfig final : public Config {
+private:
+    enum class MinerFundStatus {
+        Unset = 0,
+        Enabled = 1,
+        Disabled = 0,
+    };
+
 public:
     GlobalConfig();
     bool SetMaxBlockSize(uint64_t maxBlockSize) override;
@@ -40,8 +50,12 @@ public:
     void SetExcessUTXOCharge(Amount) override;
     Amount GetExcessUTXOCharge() const override;
 
+    void SetEnableMinerFund(bool) override;
+    bool EnableMinerFund() const override;
+
 private:
     bool useCashAddr;
+    MinerFundStatus enableMinerFund;
     Amount excessUTXOCharge;
 
     /** The largest block size this node will accept. */
@@ -66,8 +80,12 @@ public:
     void SetExcessUTXOCharge(Amount amt) override {}
     Amount GetExcessUTXOCharge() const override { return Amount::zero(); }
 
+    void SetEnableMinerFund(bool) override;
+    bool EnableMinerFund() const override;
+
 private:
     std::unique_ptr<CChainParams> chainParams;
+    bool enableMinerFund;
 };
 
 // Temporary woraround.
diff --git a/src/init.cpp b/src/init.cpp
index 193618594..79f1e822a 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -2802,6 +2802,11 @@ bool AppInitMain(Config &config, RPCServer &rpcServer,
     // We do this by default to avoid confusion with BTC addresses.
     config.SetCashAddrEncoding(args.GetBoolArg("-usecashaddr", true));
 
+    // Hidden Flag: Enable validation and addition of miner funding addresses
+    // for testing purposes
+    config.SetEnableMinerFund(args.GetBoolArg(
+        "-enableminerfund", chainparams.GetConsensus().enableMinerFund));
+
     // Step 8: load indexers
     if (args.GetBoolArg("-txindex", DEFAULT_TXINDEX)) {
         g_txindex = std::make_unique<TxIndex>(nTxIndexCache, false, fReindex);
diff --git a/src/miner.cpp b/src/miner.cpp
index 28a24216e..73992bdb8 100644
--- a/src/miner.cpp
+++ b/src/miner.cpp
@@ -67,6 +67,7 @@ BlockAssembler::BlockAssembler(const CChainParams &params,
                                const Options &options)
     : chainParams(params), m_mempool(mempool) {
     blockMinFeeRate = options.blockMinFeeRate;
+    enableMinerFund = options.enableMinerFund;
     // Limit size to between 1K and options.nExcessiveBlockSize -1K for sanity:
     nMaxGeneratedBlockSize = std::max<uint64_t>(
         1000, std::min<uint64_t>(options.nExcessiveBlockSize - 1000,
@@ -100,6 +101,8 @@ static BlockAssembler::Options DefaultOptions(const Config &config) {
         options.blockMinFeeRate = CFeeRate(n);
     }
 
+    options.enableMinerFund = config.EnableMinerFund();
+
     return options;
 }
 
@@ -209,8 +212,8 @@ BlockAssembler::CreateNewBlock(const CScript &scriptPubKeyIn) {
         amountFeeReward + GetBlockSubsidy(pblock->nBits, consensusParams);
 
     const Amount coinbaseValue = coinbaseTx.vout[1].nValue;
-    const std::vector<CTxOut> requiredOutputs =
-        GetMinerFundRequiredOutputs(consensusParams, pindexPrev, coinbaseValue);
+    const std::vector<CTxOut> requiredOutputs = GetMinerFundRequiredOutputs(
+        consensusParams, enableMinerFund, pindexPrev, coinbaseValue);
     for (const CTxOut &requiredOutput : requiredOutputs) {
         coinbaseTx.vout[1].nValue -= requiredOutput.nValue;
         coinbaseTx.vout.push_back(requiredOutput);
@@ -239,6 +242,7 @@ BlockAssembler::CreateNewBlock(const CScript &scriptPubKeyIn) {
     BlockValidationState state;
     if (!TestBlockValidity(state, chainParams, *pblock, pindexPrev,
                            BlockValidationOptions(nMaxGeneratedBlockSize)
+                               .withMinerFund(enableMinerFund)
                                .withCheckPoW(false)
                                .withCheckMerkleRoot(false))) {
         throw std::runtime_error(strprintf("%s: TestBlockValidity failed: %s",
diff --git a/src/miner.h b/src/miner.h
index 240f7324a..a582750d5 100644
--- a/src/miner.h
+++ b/src/miner.h
@@ -161,6 +161,7 @@ private:
     const CChainParams &chainParams;
 
     const CTxMemPool &m_mempool;
+    bool enableMinerFund;
 
 public:
     struct Options {
@@ -168,6 +169,7 @@ public:
         uint64_t nExcessiveBlockSize;
         uint64_t nMaxGeneratedBlockSize;
         CFeeRate blockMinFeeRate;
+        bool enableMinerFund;
     };
 
     BlockAssembler(const Config &config, const CTxMemPool &mempool);
diff --git a/src/minerfund.cpp b/src/minerfund.cpp
index e68ff8b29..9a9e473f3 100644
--- a/src/minerfund.cpp
+++ b/src/minerfund.cpp
@@ -25,12 +25,12 @@ static const CTxOut BurnOutput(const Amount amount) {
 }
 
 std::vector<CTxOut> GetMinerFundRequiredOutputs(const Consensus::Params &params,
+                                                const bool enableMinerFund,
                                                 const CBlockIndex *pindexPrev,
                                                 const Amount &blockReward) {
-    if (!gArgs.GetBoolArg("-enableminerfund", params.enableMinerFund)) {
+    if (!enableMinerFund) {
         return {};
     }
-
     const Amount shareAmount = blockReward / 26;
 
     return {
diff --git a/src/minerfund.h b/src/minerfund.h
index 51a64337c..0c71148f7 100644
--- a/src/minerfund.h
+++ b/src/minerfund.h
@@ -18,6 +18,7 @@ struct Params;
 }
 
 std::vector<CTxOut> GetMinerFundRequiredOutputs(const Consensus::Params &params,
+                                                const bool enableMinerFund,
                                                 const CBlockIndex *pindexPrev,
                                                 const Amount &blockReward);
 
diff --git a/src/rpc/mining.cpp b/src/rpc/mining.cpp
index cd9973275..c982927c5 100644
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -953,8 +953,9 @@ static UniValue getblocktemplate(const Config &config,
     UniValue aux(UniValue::VOBJ);
 
     UniValue requiredOutputsList(UniValue::VARR);
-    for (auto requiredOutput : GetMinerFundRequiredOutputs(
-             consensusParams, pindexPrev, coinbasevalue)) {
+    for (auto requiredOutput :
+         GetMinerFundRequiredOutputs(consensusParams, config.EnableMinerFund(),
+                                     pindexPrev, coinbasevalue)) {
         UniValue requiredOutputObj(UniValue::VOBJ);
         requiredOutputObj.pushKV("scriptPubKey",
                                  HexStr(requiredOutput.scriptPubKey));
diff --git a/src/test/blockfilter_index_tests.cpp b/src/test/blockfilter_index_tests.cpp
index 1478958b3..21884cc2d 100644
--- a/src/test/blockfilter_index_tests.cpp
+++ b/src/test/blockfilter_index_tests.cpp
@@ -67,7 +67,7 @@ static bool CheckFilterLookups(BlockFilterIndex &filter_index,
 CBlock BuildChainTestingSetup::CreateBlock(
     const CBlockIndex *prev, const std::vector<CMutableTransaction> &txns,
     const CScript &scriptPubKey) {
-    const Config &config = GetConfig();
+    const Config &config = DummyConfig();
     std::unique_ptr<CBlockTemplate> pblocktemplate =
         BlockAssembler(config, *m_node.mempool).CreateNewBlock(scriptPubKey);
     CBlock &block = pblocktemplate->block;
diff --git a/src/test/miner_tests.cpp b/src/test/miner_tests.cpp
index b7b50a861..eca19f531 100644
--- a/src/test/miner_tests.cpp
+++ b/src/test/miner_tests.cpp
@@ -41,15 +41,14 @@
 
 namespace miner_tests {
 struct MinerTestingSetup : public TestingSetup {
-    void TestPackageSelection(const CChainParams &chainparams,
-                              const CScript &scriptPubKey,
+    void TestPackageSelection(const Config &config, const CScript &scriptPubKey,
                               const std::vector<CTransactionRef> &txFirst)
         EXCLUSIVE_LOCKS_REQUIRED(::cs_main, m_node.mempool->cs);
     bool TestSequenceLocks(const CTransaction &tx, int flags)
         EXCLUSIVE_LOCKS_REQUIRED(::cs_main, m_node.mempool->cs) {
         return CheckSequenceLocks(*m_node.mempool, tx, flags);
     }
-    BlockAssembler AssemblerForTest(const CChainParams &params);
+    BlockAssembler AssemblerForTest(const Config &config);
 };
 } // namespace miner_tests
 
@@ -57,9 +56,11 @@ BOOST_FIXTURE_TEST_SUITE(miner_tests, MinerTestingSetup)
 
 static CFeeRate blockMinFeeRate = CFeeRate(DEFAULT_BLOCK_MIN_TX_FEE_PER_KB);
 
-BlockAssembler MinerTestingSetup::AssemblerForTest(const CChainParams &params) {
+BlockAssembler MinerTestingSetup::AssemblerForTest(const Config &config) {
+    const CChainParams &params = config.GetChainParams();
     BlockAssembler::Options options;
     options.blockMinFeeRate = blockMinFeeRate;
+    options.enableMinerFund = config.EnableMinerFund();
     return BlockAssembler(params, *m_node.mempool, options);
 }
 
@@ -106,7 +107,7 @@ static CBlockIndex CreateBlockIndex(int nHeight)
 // Implemented as an additional function, rather than a separate test case,
 // to allow reusing the blockchain created in CreateNewBlock_validity.
 void MinerTestingSetup::TestPackageSelection(
-    const CChainParams &chainparams, const CScript &scriptPubKey,
+    const Config &config, const CScript &scriptPubKey,
     const std::vector<CTransactionRef> &txFirst) {
     // Test the ancestor feerate transaction selection.
     TestMemPoolEntryHelper entry;
@@ -157,7 +158,7 @@ void MinerTestingSetup::TestPackageSelection(
                                      .FromTx(tx));
 
     std::unique_ptr<CBlockTemplate> pblocktemplate =
-        AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey);
+        AssemblerForTest(config).CreateNewBlock(scriptPubKey);
     BOOST_CHECK_EQUAL(pblocktemplate->block.vtx.size(), 4);
     BOOST_CHECK_EQUAL(pblocktemplate->block.vtx[1]->GetId(), parentTxId);
     BOOST_CHECK_EQUAL(pblocktemplate->block.vtx[2]->GetId(), highFeeTxId);
@@ -180,7 +181,7 @@ void MinerTestingSetup::TestPackageSelection(
     tx.vout[0].nValue = freeTxAmount - feeToUse;
     TxId lowFeeTxId = tx.GetId();
     m_node.mempool->addUnchecked(entry.Fee(feeToUse).FromTx(tx));
-    pblocktemplate = AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey);
+    pblocktemplate = AssemblerForTest(config).CreateNewBlock(scriptPubKey);
     // Verify that the free tx and the low fee tx didn't get selected.
     for (const auto &txn : pblocktemplate->block.vtx) {
         BOOST_CHECK(txn->GetId() != freeTxId);
@@ -196,7 +197,7 @@ void MinerTestingSetup::TestPackageSelection(
     tx.vout[0].nValue -= 2 * SATOSHI;
     lowFeeTxId = tx.GetId();
     m_node.mempool->addUnchecked(entry.Fee(feeToUse + 2 * SATOSHI).FromTx(tx));
-    pblocktemplate = AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey);
+    pblocktemplate = AssemblerForTest(config).CreateNewBlock(scriptPubKey);
     BOOST_CHECK_EQUAL(pblocktemplate->block.vtx.size(), 6);
     BOOST_CHECK_EQUAL(pblocktemplate->block.vtx[1]->GetId(), freeTxId);
     BOOST_CHECK_EQUAL(pblocktemplate->block.vtx[2]->GetId(), lowFeeTxId);
@@ -226,7 +227,7 @@ void MinerTestingSetup::TestPackageSelection(
     TxId lowFeeTxId2 = tx.GetId();
     m_node.mempool->addUnchecked(
         entry.Fee(feeToUse).SpendsCoinbase(false).FromTx(tx));
-    pblocktemplate = AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey);
+    pblocktemplate = AssemblerForTest(config).CreateNewBlock(scriptPubKey);
 
     // Verify that this tx isn't selected.
     for (const auto &txn : pblocktemplate->block.vtx) {
@@ -240,7 +241,7 @@ void MinerTestingSetup::TestPackageSelection(
     // 10k satoshi fee.
     tx.vout[0].nValue = (100000000 - 10000) * SATOSHI;
     m_node.mempool->addUnchecked(entry.Fee(10000 * SATOSHI).FromTx(tx));
-    pblocktemplate = AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey);
+    pblocktemplate = AssemblerForTest(config).CreateNewBlock(scriptPubKey);
     BOOST_CHECK_EQUAL(pblocktemplate->block.vtx[3]->GetId(), lowFeeTxId2);
 }
 
@@ -282,6 +283,8 @@ BOOST_AUTO_TEST_CASE(CheckCoinbase_EB) {
 BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) {
     // Note that by default, these tests run with size accounting enabled.
     GlobalConfig config;
+    // Don't check this functionality in this test.
+    config.SetEnableMinerFund(false);
     const CChainParams &chainparams = config.GetChainParams();
     CScript scriptPubKey =
         CScript() << ParseHex("04678afdb0fe5548271967f1a67130b7105cd6a828e03909"
@@ -299,7 +302,7 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) {
 
     // Simple block creation, nothing special yet:
     BOOST_CHECK(pblocktemplate =
-                    AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey));
+                    AssemblerForTest(config).CreateNewBlock(scriptPubKey));
 
     // We can't make transactions until we have inputs.
     // Therefore, load 110 blocks :)
@@ -354,7 +357,8 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) {
             Amount minerReward =
                 GetBlockSubsidy(pblock->nBits, chainparams.GetConsensus());
             std::vector<CTxOut> requiredOutputs = GetMinerFundRequiredOutputs(
-                chainparams.GetConsensus(), ::ChainActive().Tip(), minerReward);
+                chainparams.GetConsensus(), config.EnableMinerFund(),
+                ::ChainActive().Tip(), minerReward);
             for (size_t i = 0; i < requiredOutputs.size(); ++i) {
                 txCoinbase.vout[i + 2] = requiredOutputs[i];
                 minerReward -= requiredOutputs[i].nValue;
@@ -397,7 +401,7 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) {
 
     // Just to make sure we can still make simple blocks.
     BOOST_CHECK(pblocktemplate =
-                    AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey));
+                    AssemblerForTest(config).CreateNewBlock(scriptPubKey));
 
     std::vector<uint8_t> padData(100);
 
@@ -428,14 +432,14 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) {
     }
 
     BOOST_CHECK(pblocktemplate =
-                    AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey));
+                    AssemblerForTest(config).CreateNewBlock(scriptPubKey));
     m_node.mempool->clear();
 
     // Orphan in mempool, template creation fails.
     m_node.mempool->addUnchecked(entry.Fee(LOWFEE).Time(GetTime()).FromTx(tx));
-    BOOST_CHECK_EXCEPTION(
-        AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey),
-        std::runtime_error, HasReason("bad-txns-inputs-missingorspent"));
+    BOOST_CHECK_EXCEPTION(AssemblerForTest(config).CreateNewBlock(scriptPubKey),
+                          std::runtime_error,
+                          HasReason("bad-txns-inputs-missingorspent"));
     m_node.mempool->clear();
 
     // Child with higher priority than parent.
@@ -456,7 +460,7 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) {
     m_node.mempool->addUnchecked(
         entry.Fee(HIGHERFEE).Time(GetTime()).SpendsCoinbase(true).FromTx(tx));
     BOOST_CHECK(pblocktemplate =
-                    AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey));
+                    AssemblerForTest(config).CreateNewBlock(scriptPubKey));
     m_node.mempool->clear();
 
     // Coinbase in mempool, template creation fails.
@@ -469,9 +473,8 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) {
     m_node.mempool->addUnchecked(
         entry.Fee(LOWFEE).Time(GetTime()).SpendsCoinbase(false).FromTx(tx));
     // Should throw bad-tx-coinbase
-    BOOST_CHECK_EXCEPTION(
-        AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey),
-        std::runtime_error, HasReason("bad-tx-coinbase"));
+    BOOST_CHECK_EXCEPTION(AssemblerForTest(config).CreateNewBlock(scriptPubKey),
+                          std::runtime_error, HasReason("bad-tx-coinbase"));
     m_node.mempool->clear();
 
     // Double spend txn pair in mempool, template creation fails.
@@ -489,9 +492,9 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) {
     txid = tx.GetId();
     m_node.mempool->addUnchecked(
         entry.Fee(HIGHFEE).Time(GetTime()).SpendsCoinbase(true).FromTx(tx));
-    BOOST_CHECK_EXCEPTION(
-        AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey),
-        std::runtime_error, HasReason("bad-txns-inputs-missingorspent"));
+    BOOST_CHECK_EXCEPTION(AssemblerForTest(config).CreateNewBlock(scriptPubKey),
+                          std::runtime_error,
+                          HasReason("bad-txns-inputs-missingorspent"));
     m_node.mempool->clear();
 
     // Subsidy changing.
@@ -513,7 +516,7 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) {
         ::ChainActive().SetTip(next);
     }
     BOOST_CHECK(pblocktemplate =
-                    AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey));
+                    AssemblerForTest(config).CreateNewBlock(scriptPubKey));
     // Extend to a 210000-long block chain.
     while (::ChainActive().Tip()->nHeight < 210000) {
         CBlockIndex *prev = ::ChainActive().Tip();
@@ -531,7 +534,7 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) {
     }
 
     BOOST_CHECK(pblocktemplate =
-                    AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey));
+                    AssemblerForTest(config).CreateNewBlock(scriptPubKey));
 
     // Invalid p2sh txn in mempool, template creation fails
     tx.vin[0].prevout = COutPoint(txFirst[0]->GetId(), 1);
@@ -550,9 +553,8 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) {
     m_node.mempool->addUnchecked(
         entry.Fee(LOWFEE).Time(GetTime()).SpendsCoinbase(false).FromTx(tx));
     // Should throw blk-bad-inputs
-    BOOST_CHECK_EXCEPTION(
-        AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey),
-        std::runtime_error, HasReason("blk-bad-inputs"));
+    BOOST_CHECK_EXCEPTION(AssemblerForTest(config).CreateNewBlock(scriptPubKey),
+                          std::runtime_error, HasReason("blk-bad-inputs"));
     m_node.mempool->clear();
 
     // Delete the dummy blocks again.
@@ -741,7 +743,7 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) {
     // Sequence locks fail.
     BOOST_CHECK(!TestSequenceLocks(CTransaction(tx), flags));
 
-    pblocktemplate = AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey);
+    pblocktemplate = AssemblerForTest(config).CreateNewBlock(scriptPubKey);
     BOOST_CHECK(pblocktemplate);
 
     // None of the of the absolute height/time locked tx should have made it
@@ -762,14 +764,14 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity) {
     SetMockTime(::ChainActive().Tip()->GetMedianTimePast() + 1);
 
     BOOST_CHECK(pblocktemplate =
-                    AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey));
+                    AssemblerForTest(config).CreateNewBlock(scriptPubKey));
     BOOST_CHECK_EQUAL(pblocktemplate->block.vtx.size(), 5UL);
 
     ::ChainActive().Tip()->nHeight--;
     SetMockTime(0);
     m_node.mempool->clear();
 
-    TestPackageSelection(chainparams, scriptPubKey, txFirst);
+    TestPackageSelection(config, scriptPubKey, txFirst);
 
     fCheckpointsEnabled = true;
 }
diff --git a/src/validation.cpp b/src/validation.cpp
index a71f6fd82..56f853b7a 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -148,7 +148,7 @@ std::set<int> setDirtyFileInfo;
 
 BlockValidationOptions::BlockValidationOptions(const Config &config)
     : excessiveBlockSize(config.GetMaxBlockSize()), checkPoW(true),
-      checkMerkleRoot(true) {}
+      checkMerkleRoot(true), enableMinerFund(config.EnableMinerFund()) {}
 
 CBlockIndex *LookupBlockIndex(const BlockHash &hash) {
     AssertLockHeld(cs_main);
@@ -1902,7 +1902,8 @@ bool CChainState::ConnectBlock(const CBlock &block, BlockValidationState &state,
     }
 
     const std::vector<CTxOut> requiredOutputs = GetMinerFundRequiredOutputs(
-        consensusParams, pindex->pprev, blockReward);
+        consensusParams, options.shouldValidateMinerFund(), pindex->pprev,
+        blockReward);
 
     if (!requiredOutputs.empty()) {
         auto nextRequiredOutput = requiredOutputs.begin();
diff --git a/src/validation.h b/src/validation.h
index d15d5be3a..d016bd2ce 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -14,8 +14,8 @@
 #include <amount.h>
 #include <blockfileinfo.h>
 #include <blockindexworkcomparator.h>
-#include <coins.h>
 #include <chain.h>
+#include <coins.h>
 #include <consensus/consensus.h>
 #include <disconnectresult.h>
 #include <flatfile.h>
@@ -183,15 +183,18 @@ private:
     uint64_t excessiveBlockSize;
     bool checkPoW : 1;
     bool checkMerkleRoot : 1;
+    bool enableMinerFund : 1;
 
 public:
     // Do full validation by default
     explicit BlockValidationOptions(const Config &config);
     explicit BlockValidationOptions(uint64_t _excessiveBlockSize,
                                     bool _checkPow = true,
-                                    bool _checkMerkleRoot = true)
+                                    bool _checkMerkleRoot = true,
+                                    bool _enableMinerFund = true)
         : excessiveBlockSize(_excessiveBlockSize), checkPoW(_checkPow),
-          checkMerkleRoot(_checkMerkleRoot) {}
+          checkMerkleRoot(_checkMerkleRoot), enableMinerFund(_enableMinerFund) {
+    }
 
     BlockValidationOptions withCheckPoW(bool _checkPoW = true) const {
         BlockValidationOptions ret = *this;
@@ -206,9 +209,16 @@ public:
         return ret;
     }
 
+    BlockValidationOptions withMinerFund(bool _enableMinerFund = true) const {
+        BlockValidationOptions ret = *this;
+        ret.enableMinerFund = _enableMinerFund;
+        return ret;
+    }
+
     bool shouldValidatePoW() const { return checkPoW; }
     bool shouldValidateMerkleRoot() const { return checkMerkleRoot; }
     uint64_t getExcessiveBlockSize() const { return excessiveBlockSize; }
+    bool shouldValidateMinerFund() const { return enableMinerFund; }
 };
 
 /**
-- 
2.39.5 (Apple Git-154)

